feat(upgrade): introduce tracing as an optional unstable feature

RamziA961 opened the pull request:
This change allow users to opt out of tracing via the `tracing` crate by adding tracing as an optional feature. This change is part of the effort, outlined in #2874, to reach hyper 1.0.

Closes #3319
BREAKING CHANGES: tracing is disabled by default and requires users to opt in to revert to previous behavior.

seanmonstar commented:
Thanks for taking this on!

We need to make it unstable similar to the FFI feature, with a check in `lib.rs` that a `cfg(hyper_unstable_tracing)` was also passed, or trigger a compiler error. And then document it in the unstable features section of the same file.

Separately, one possible way I think this could be done is making an internal module with trace and debug macros, and inside those out the cfg checks. It's not a big deal, it just might be easier.

RamziA961 commented:
Thanks for the feedback. An internal module would definitely be a cleaner approach, and would help with unused variable warnings. I'll get started on that.

RamziA961 commented:
I think a quick overview of the changes I made might be helpful in review the PR, so here goes:

 - I introduced a module called `trace`, which essentially exposes macros that act as wrappers around `tracing`'s macros but handle conditional compilation internally. The caveat is `tracing` has `span` macros aren't so easily replaceable. Their uses throughout the codebase are sparse and all uses comprised creating  a `Span` and the calling the method `enter()`. Both the `Span` and `Entered` instances need to be bound to a variable so that they are not dropped. The only approach that I could think of the preserves the conciseness of the original syntax is the following but it is greatly restrictive:

```rust
macro_rules! debug_span {
    ($($arg:tt)*) => {
        #[cfg(feature = "tracing")]
        {
            #[cfg(not(hyper_unstable_tracing))]
            compile_error!(
                "\
                The `tracing` feature is unstable, and requires the \
                `RUSTFLAGS='--cfg hyper_unstable_tracing'` environment variable to be set.\
            "
            );
            let span = tracing::debug_span!($($arg)+);
            let _ = span.enter();
        }
    }
}
```

- I replaced uses of `tracing`'s macros with the new internal `trace` module's macro. In most cases, it simply required deleting `use tracing::{x, y, z}`, except when the macros were used as part of an expression, e.g., `something.map_err(|e| tracing::debug!("{e:?}"))`.

- Some variables had to be prefix with an underscore since unused variables warnings are emitted when compiling with `tracing` disabled.

- To ensure that the codes compiles when the `tracing` feature is off, the `trace` module must always be declared in `lib` so that the compiler can expand invocations of its macros into nothing.

- I introduced a new feature `tracing` and a new cfg flag `--hyper_unstable_tracing`. The feature must be enabled and the flag must be passed to `rustc` for tracing to be enabled, otherwise a compiler error is thrown in similar fashion to `ffi`. This is the cause of the failing CI test.

- Documentation for the module, feature, and cfg flag are provided.

I think that covers it.

seanmonstar commented on src/lib.rs:
```diff
@@ -67,6 +67,11 @@ pub use crate::error::{Error, Result};
 mod cfg;
+
+#[macro_use]
+#[cfg_attr(docsrs, doc(all(feature = "tracing"), hyper_unstable_tracing))]
```
Is this attribute required? Since it's not public, the module won't be rendered on docs.rs.

seanmonstar commented on src/trace.rs:
```diff
@@ -0,0 +1,212 @@
+            "
+            );
+            let span = tracing::debug_span!($($arg)+);
+            let _ = span.enter();
```
I believe this will drop the enter guard immediately, since the block ends on the next line. You could change this to return `span.entered()`, and then the usage would be `let _s = debug_span!(blah);`.

hawkw replied:
yeah, that's correct. we should either have the macro return an un-entered span that's entered at the callsite, or change this to 
```suggestion
            let _enter = span.enter();
```
or similar.

RamziA961 replied:
The issue with `let _s = debug_span!(blah);` is that the compiler does not support removing an expression. When tracing is off, this effectively becomes `let _s = ;`. 

Similarly, `let _enter  = span.enter()` is also dropped at the end of the block.

The following seems to work:
```rust
use tracing_subscriber::prelude::*;

macro_rules! test {
    ($($arg:tt)+) => {
        #[cfg(debug_assertions)]
        { 
            let _sp = tracing::error_span!($($arg)+);
            _sp.entered()
        }
    };
}

fn main() {
    let stdout = tracing_subscriber::fmt::layer().pretty();

    tracing_subscriber::registry()
        .with(
            stdout
                .with_span_events(tracing_subscriber::fmt::format::FmtSpan::EXIT)
        )
        .init();
     
    let _e = { test!("asd", value = 32 ) };
    
    tracing::error!("test one");
    tracing::error!("test two");
}
```

Should I move forward with this approach?

seanmonstar commented on src/trace.rs:
```diff
@@ -0,0 +1,212 @@
+//! crate. However, as the contents of these macros are conditionally compiled, they are effectively trimmed
+//! during inline expansion when the `tracing` feature is disabled.
+//!
+//! # Unstable
```
This makes me realize that in `lib.rs`, there isn't a section about unstable features. We should move the essence of this there (without describing what the internals of this module does).

seanmonstar reviewed (commented):
Thanks so much, the work is very thorough, and your write up is spectacular.

hawkw commented on Cargo.toml:
```diff
@@ -29,14 +29,14 @@ http-body-util = { version = "=0.1.0-rc.3", optional = true }
 
 httpdate = { version = "1.0", optional = true }
 libc = { version = "0.2", optional = true }
+tracing = { version = "0.1", default-features = false, features = ["std"], optional = true }
```
is `tracing`'s "std" feature actually required here? i think we should be fine without enabling it (although it doesn't matter, since hyper doesn't support no-std...)

hawkw commented on src/trace.rs:
```diff
@@ -0,0 +1,212 @@
+            "
+            );
+            let span = tracing::error_span!($($arg)+);
+            let _ = span.enter();
```
this also drops the span immediately and will need to be changed as discussed above.

hawkw commented on src/trace.rs:
```diff
@@ -0,0 +1,212 @@
+            "
+            );
+            let span = tracing::info_span!($($arg)+);
+            let _ = span.enter();
```
same issue as above.

hawkw commented on src/trace.rs:
```diff
@@ -0,0 +1,212 @@
+            "
+            );
+            let span = tracing::trace_span!($($arg)+);
+            let _ = span.enter();
```
and here

hawkw commented on src/proto/h2/client.rs:
```diff
@@ -243,7 +242,9 @@ where
         }
-        polled.map_err(|e| debug!("connection error: {}", e))
+        polled.map_err(|_e| {
+            debug!("connection error: {}", _e);
```
while we're here, we may want to change these events to record the error as a structured field, like
```suggestion
            debug!(error = %_e, "connection error");
```
this isn't a big deal, but it could be worth doing, and we're changing all the existing tracing macros anyway...

hawkw commented on src/trace.rs:
```diff
@@ -0,0 +1,212 @@
+    ($($arg:tt)+) => {
+        #[cfg(feature = "tracing")]
+        {
+            #[cfg(not(hyper_unstable_tracing))]
+            compile_error!(
+                "\
+                The `tracing` feature is unstable, and requires the \
+                `RUSTFLAGS='--cfg hyper_unstable_tracing'` environment variable to be set.\
+            "
+            );
```
this will result in the same compile error being emitted for *every* use of a tracing macro, if the unstable cfg is unset. this means there will be a giant pile of the same error in the rustc output. i wonder if it would be nicer to instead just stick the compile error in this module, so that we generate a single error if the feature flag is enabled without the unstable cfg?

RamziA961 replied:
I agree, I will make the change.

hawkw commented on src/upgrade.rs:
```diff
@@ -233,6 +231,7 @@ impl fmt::Debug for OnUpgrade {
 #[cfg(any(feature = "http1", feature = "http2"))]
 impl Pending {
     pub(super) fn fulfill(self, upgraded: Upgraded) {
+        #[cfg(any(feature = "http1", feature = "http2"))]
```
why was it necessary to add these additional cfg attributes? the entire impl block is only enabled with `#[cfg(any(feature = "http1", feature = "http2"))]`, so AFAICT, this shouldn't be necessary?

RamziA961 replied:
Good catch. Thank you for pointing this out.

seanmonstar commented:
All the things look good to me, thanks! Good job! You could include the change to the CI features check in this PR, so that it merges with the relevant code. Also, looks like there's merge conflicts.
